framework/src/functormaterials/FunctorSmoother.C

```diff
@@ -0,0 +1,225 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FunctorSmoother.h"
+#include "MooseUtils.h"
+#include "RelationshipManager.h"
+
+#include "metaphysicl/raw_type.h"
+
+registerMooseObject("MooseApp", FunctorSmoother);
+
+template <typename T>
+InputParameters
+FunctorSmootherTempl<T>::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription("Creates smoother functor(s) using various averaging techniques");
+  params.addParam<std::vector<MooseFunctorName>>("functors_in",
+                                                 "The name(s) of the functors to smooth");
+  params.addParam<std::vector<MooseFunctorName>>("functors_out",
+                                                 "The name(s) of the smooth output functors");
+  MooseEnum smoothing_techniques("face_average layered_elem_average remove_checkerboard");
+  params.addParam<MooseEnum>(
+      "smoothing_technique", smoothing_techniques, "How to smooth the functor");
+
+  return params;
+}
+
+template <typename T>
+FunctorSmootherTempl<T>::FunctorSmootherTempl(const InputParameters & parameters)
+  : FunctorMaterial(parameters),
+    _functors_in(getParam<std::vector<MooseFunctorName>>("functors_in")),
+    _functors_out(getParam<std::vector<MooseFunctorName>>("functors_out")),
+    _smoothing_technique(getParam<MooseEnum>("smoothing_technique"))
+{
+  if (_tid == 0)
+  {
+    // We need one layer of ghosting at least to get neighbor values
+    auto & factory = _app.getFactory();
+    auto rm_params = factory.getValidParams("ElementSideNeighborLayers");
+    rm_params.template set<std::string>("for_whom") = name();
+    rm_params.template set<MooseMesh *>("mesh") = &const_cast<MooseMesh &>(_mesh);
+    rm_params.template set<Moose::RelationshipManagerType>("rm_type") =
+        Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+        Moose::RelationshipManagerType::COUPLING;
+    rm_params.template set<unsigned short>("layers") = 2 + (_smoothing_technique == FACE_AVERAGE);
```

Comment:
I don't understand why there needs to be more ghosting for the face averaging technique

Reply:
oh this is residue from testing

Reply:
looks like we pass testing, lemme remove that